doc: Matter protocol: add Group Communication (#19954)
nrfconnect/sdk-nrf: doc/nrf/protocols/matter/overview/group_communication.rst

* Capacity Management: The cluster handles the limitations on how many groups a device can be a member of, ensuring that device capabilities are not exceeded.

Group Key Management Cluster
============================

The Group Key Management Cluster is crucial for the security aspect of Group Communication.
It manages the encryption keys that are used to secure multicast messages sent to groups.
This ensures that only the members of a group can read the messages intended for that group, maintaining confidentiality and integrity.
A cluster should be supported for both the device that is an transmitter and the reciver of Group Communication.

The following are the key functions of a Group Key Management Cluster:

* Key Set Creation and Distribution: This function involves generating and distributing the keys that will be used to encrypt and decrypt messages sent to the group.
  The distribution must be secure to prevent interception and unauthorized access.
* Key Storage and Handling: Secure storage and handling of keys within devices are managed to prevent unauthorized access and ensure that keys are available when needed for encryption and decryption processes.

> Example use cases

Reviewer: Can we make a example use cases as tabs?
With following Sections:
* Nodes? - will define what samples, testers are used within the use case
* Precondition
* Steps

## Example for USE CASE 1:
### Nodes
* chip-tool -  ( https://github.com/nrfconnect/sdk-connectedhomeip/releases/tag/v2.9.0)
* light bulb - link to light bulb readme
### Precondition
Light Bulb should be commissioned to the same fabric as chip-tool
### Steps 
Steps defined in the use case

## Example for USE CASE 2:
### Nodes
* chip-tool -  ( https://github.com/nrfconnect/sdk-connectedhomeip/releases/tag/v2.9.0)
* light bulb - link to light bulb readme
* light  switch - link to light switch readme
### Precondition
Light Bulb and Light Switch should be commissioned to the same fabric as chip-tool
### Steps 
Steps defined in the use case

Author: Done.

*****************

> The following usecases demonstrate how you can manage the Group Communication settings using the ``chip-tool``.

Reviewer: maybe we can link here also the documentation to samples used in that use-cases:
 * first scenario: chip-tool (as described)  - group communication initiator and samples/matter/light_bulb
 * second scenario: samples: samples/matter/light_switch  - group communication initiator and samples/matter/light_bulb

Author: Done.

Group Communication triggered from the controller
=================================================

Check first your settings, keysets and groups.

* To see the Group Cluster and Group Key Management Cluster settings you can manage, use the following command:

>   .. code-block:: none

Reviewer: maybe this part with explaining the chip-tool commands group settings move out of the use case scope, and move to separate e.g. chip-tool group communication configuration settings or something different

Author: Done.

   ./chip-tool groupkeymanagement key-set-write '{"groupKeySetID":"0xabcd","groupKeySecurityPolicy":0,"epochKey0":"0123456789abcdef0123456789abcdef","epochStartTime0":"3330000","epochKey1":"0123456789abcdef0123456789abcdee","epochStartTime1":"3330001","epochKey2":"0123456789abcdef0123456789abcded","epochStartTime2":"3330002"}' 1 0

#. Set the keyset to the group map::

   ./chip-tool groupkeymanagement write group-key-map '[{"groupId":"0xdcba","groupKeySetID":"0xabcd","fabricIndex":"1"}]' 1 0

#. Add the group on the application endpoint (endpoint 1 includes the on/off cluster)::

   ./chip-tool groups add-group 0xdcba Custom_Group 1 1

#. Set the ACL to permit the Group Communication from a specific group ID::

   ./chip-tool accesscontrol write acl '[{"fabricIndex":"1","privilege":"5","authMode":"2","subjects":["112233"],"targets":null},{"fabricIndex":"1","privilege":"3","authMode":"3","subjects":["0xdcba"],"targets":[{"cluster":null,"endpoint":"1","deviceType":null}]}]' 1 0

#. Send a command to toggle on/off on the application endpoint (endpoint 1 includes the onoff cluster)::

   ./chip-tool onoff toggle 0xffffffffffffdcba 1

> **LED2** on the Light Bulb should now change the state.

Reviewer: It is LED 2 for nrf53/nrf52,  for nrf54l15 is LED1 (because numeration is different on DK)

#. Set the same group settings on the on/off server (Light Bulb)::

   ./chip-tool groupkeymanagement key-set-write '{"groupKeySetID":"258","groupKeySecurityPolicy":0,"epochKey0":"a0a1a2a3a4a5a6a7a8a9aaabacad7531","epochStartTime0":"1110000","epochKey1":"b0b1b2b3b4b5b6b7b8b9babbbcbd7531","epochStartTime1":"1110001","epochKey2":"c0c1c2c3c4c5c6c7c8c9cacbcccd7531","epochStartTime2":"1110002"}' 2 0

#. Set the keyset to the group map on the on/off server (Light Bulb)::

   ./chip-tool groupkeymanagement write group-key-map '[{"groupId":"30001","groupKeySetID":"258","fabricIndex":"1"}]' 2 0

#. Add the group to the application endpoint (endpoint 1 includes the on/off cluster)::

   ./chip-tool groups add-group 30001 Test_Group_30001 2 1

#. Set the binding to the on/off client (Light Switch)::

   ./chip-tool binding write binding '[{"fabricIndex":"1","group":"30001"}]' 1 1

#. Set ACL on the on/off server (Light Bulb)::

   ./chip-tool accesscontrol write acl '[{"fabricIndex":1,"privilege":5,"authMode":2,"subjects":[112233],"targets":null},{"fabricIndex":1,"privilege":3,"authMode":3,"subjects":[30001],"targets":[{"cluster":null,"endpoint":1,"deviceType":null}]}]' 2 0

> #. Click button **2** on the Light Switch to send the toggle message.

Reviewer: It is BUTTON 2 for nrf53/nrf52,  for nrf54l15 is BUTTON1 (because numeration is different on DK)

Author: Done.

> **LED2** on the Light Bulb should now change the state.

Reviewer: It is LED 2 for nrf53/nrf52,  for nrf54l15 is LED1 (because numeration is different on DK)

.. _ug_matter_group_communication:

Matter Group Communication
##########################

.. contents::
   :local:
   :depth: 2

This section describes how Group Communication works in Matter.

Group Communication in the Matter standard refers to the ability of devices within a Matter network to communicate with multiple devices simultaneously using multicast messages.
This is particularly useful for scenarios where a command needs to be sent to multiple devices at once, such as turning off all lights in a house or setting a scene.

Group setup
***********

Devices are configured to be part of one or more groups.
Each group is identified by a unique group ID.
A device can be a member of multiple groups, and a group can include multiple devices.

Group Messaging
***************

When a command needs to be sent to all devices in a group, the command initiator (a controller, a smartphone app, or another device) sends a multicast message addressed to the group ID.

> Only the devices that are members of that group will respond to or act on the message.

Reviewer: You are right, the respond, its a wrong word, it should more like `receive`.

> Only the devices that are members of that group will respond to or act on the message.

Author: Fixed.

> Only the devices that are members of that group will respond to or act on the message.

Reviewer 2: Sorry, but I didn't manage to comment before you changed it :) I'm not quite convinced to using `receive` as well. It works like this:
1. Controller sends multicast message to everyone
2. All devices receive it and see if the group id applies to them or not (sort of filtering)
3. If it's their group id, the message is processed, otherwise it's dropped.

I would either say that only the devices that are members will process and act on the message or remove the part before `or` at all and leave only act on the message.

Using Group Communication reduces network traffic because a single multicast message can reach multiple devices.

> This is more efficient than sending individual messages to each device.

Author: Please suggest some wording.

> This is more efficient than sending individual messages to each device.

Reviewer: you are totally right,
I thin we can add these limitations which Kamil mentioned.

> This is more efficient than sending individual messages to each device.

Author: Added a note. Check whether you are happy with that.

Security provisions are in place to manage who can add or remove devices from groups, preventing unauthorized control.
Common use cases include lighting control (for example, turning all lights in a room or building on or off), temperature adjustments for HVAC systems in multiple zones, and scene setting where multiple device types perform coordinated actions (such as lights dim, blinds close, temperature adjusts).

In the Matter specification, managing Group Communication effectively and securely involves specific clusters such as the Group Cluster and the Group Key Management Cluster.
These clusters play crucial roles in handling group configurations and ensuring secure communication within groups.

Group Cluster
=============

The Group Cluster is primarily responsible for managing group memberships within devices.
It allows devices to be added to or removed from groups and to query their group memberships.
This cluster is essential for organizing devices into logical sets that can receive common commands, which is fundamental for scenarios like controlling multiple lights with a single command.

> Group Clusters should be supported only for devices that may receive groupcast communication.

Reviewer 3: ```suggestion
Group Clusters should be supported only for devices that can receive groupcast communication.
```

The following are the key functions of a Group Cluster:

> * Add Group/Remove Group: These commands allow devices to be dynamically added to or removed from specific groups.
> * View Group Membership: Devices can report which groups they belong to, which is useful for controllers to make decisions about sending commands.
> * Capacity Management: The cluster handles the limitations on how many groups a device can be a member of, ensuring that device capabilities are not exceeded.

Reviewer 3: ```suggestion
* Add Group/Remove Group - These commands allow devices to be dynamically added to or removed from specific groups.
* View Group Membership - Devices can report which groups they belong to, which is useful for controllers to make decisions about sending commands.
* Capacity Management - The cluster handles the limitations on how many groups a device can be a member of, ensuring that device capabilities are not exceeded.
```

Group Key Management Cluster
============================

The Group Key Management Cluster is crucial for the security aspect of Group Communication.
It manages the encryption keys that are used to secure multicast messages sent to groups.
This ensures that only the members of a group can read the messages intended for that group, maintaining confidentiality and integrity.

> A cluster should be supported for both the device that is an transmitter and the reciver of Group Communication.

Reviewer 3: ```suggestion
A cluster should be supported for both the device that is a transmitter and the receiver of Group Communication.
```

The following are the key functions of a Group Key Management Cluster:

> * Key Set Creation and Distribution: This function involves generating and distributing the keys that will be used to encrypt and decrypt messages sent to the group.
>   The distribution must be secure to prevent interception and unauthorized access.
> * Key Storage and Handling: Secure storage and handling of keys within devices are managed to prevent unauthorized access and ensure that keys are available when needed for encryption and decryption processes.

Reviewer 3: ```suggestion
* Key Set Creation and Distribution - This function involves generating and distributing the keys that will be used to encrypt and decrypt messages sent to the group.
  The distribution must be secure to prevent interception and unauthorized access.
* Key Storage and Handling - Secure storage and handling of keys within devices are managed to prevent unauthorized access and ensure that keys are available when needed for encryption and decryption processes.
```

Example use cases
*****************

> The following usecases demonstrate how you can manage the Group Communication settings using the ``chip-tool``.

Reviewer 3: ```suggestion
The following use cases demonstrate how you can manage the Group Communication settings using the ``chip-tool``.
```

> Settings for the ``chip-tool``

Reviewer 2: Hmm, I'm wondering if linking to the tag with artifacts here is a good idea. Someone may think that under the link he or she is going to find more information about what chip tool is, while it leads to the binaries.

Author: Fixed to link to https://docs.nordicsemi.com/bundle/ncs-latest/page/matter/chip_tool_guide.html

         ./chip-tool groupsettings bind-keyset 0xdcba 0xabcd

      #. Set the keyset on the Light Bulb based on generated controller's group settings::

         ./chip-tool groupkeymanagement key-set-write '{"groupKeySetID":"0xabcd","groupKeySecurityPolicy":0,"epochKey0":"0123456789abcdef0123456789abcdef","epochStartTime0":"3330000","epochKey1":"0123456789abcdef0123456789abcdee","epochStartTime1":"3330001","epochKey2":"0123456789abcdef0123456789abcded","epochStartTime2":"3330002"}' 1 0

      #. Set the keyset to the group map::

         ./chip-tool groupkeymanagement write group-key-map '[{"groupId":"0xdcba","groupKeySetID":"0xabcd","fabricIndex":"1"}]' 1 0

      #. Add the group on the application endpoint (endpoint 1 includes the on/off cluster)::

         ./chip-tool groups add-group 0xdcba Custom_Group 1 1

      #. Set the ACL to permit the Group Communication from a specific group ID::

         ./chip-tool accesscontrol write acl '[{"fabricIndex":"1","privilege":"5","authMode":"2","subjects":["112233"],"targets":null},{"fabricIndex":"1","privilege":"3","authMode":"3","subjects":["0xdcba"],"targets":[{"cluster":null,"endpoint":"1","deviceType":null}]}]' 1 0

>       #. Send a command to toggle on/off on the application endpoint (endpoint 1 includes the onoff cluster)::

Reviewer 2: onoff->on/off? I think we used on/off convention in all other places.

Reviewer: ooh once is on/off and once onoff

Author: Fixed.

         ./chip-tool groupkeymanagement key-set-write '{"groupKeySetID":"258","groupKeySecurityPolicy":0,"epochKey0":"a0a1a2a3a4a5a6a7a8a9aaabacad7531","epochStartTime0":"1110000","epochKey1":"b0b1b2b3b4b5b6b7b8b9babbbcbd7531","epochStartTime1":"1110001","epochKey2":"c0c1c2c3c4c5c6c7c8c9cacbcccd7531","epochStartTime2":"1110002"}' 1 0

      #. Set the keyset to the group map on the on/off client (Light Switch)::

         ./chip-tool groupkeymanagement write group-key-map '[{"groupId":"30001","groupKeySetID":"258","fabricIndex":"1"}]' 1 0

      #. Set the same group settings on the on/off server (Light Bulb)::

         ./chip-tool groupkeymanagement key-set-write '{"groupKeySetID":"258","groupKeySecurityPolicy":0,"epochKey0":"a0a1a2a3a4a5a6a7a8a9aaabacad7531","epochStartTime0":"1110000","epochKey1":"b0b1b2b3b4b5b6b7b8b9babbbcbd7531","epochStartTime1":"1110001","epochKey2":"c0c1c2c3c4c5c6c7c8c9cacbcccd7531","epochStartTime2":"1110002"}' 2 0

      #. Set the keyset to the group map on the on/off server (Light Bulb)::

         ./chip-tool groupkeymanagement write group-key-map '[{"groupId":"30001","groupKeySetID":"258","fabricIndex":"1"}]' 2 0

>       #. Add the group to the application endpoint (endpoint 1 includes the on/off cluster)::

Reviewer 2: Does it apply to light bulb, light switch or both? The other steps point to which device it applies directly.

Reviewer: you are right, it should be more precisely mentioned that it should be set only on light bullb (which support group cluster on application endpoint)

Author: Please, suggest how to put this right.

Reviewer: We should change the step to `Add the group to the Light Bulb Application endpoint (endpoint 1 includes the on/off cluster)`

This is particularly useful for scenarios where a command needs to be sent to multiple devices at once, such as turning off all lights in a house or setting a scene.

Group setup
***********

Devices are configured to be part of one or more groups.
Each group is identified by a unique group ID.
A device can be a member of multiple groups, and a group can include multiple devices.

Group Messaging
***************

When a command needs to be sent to all devices in a group, the command initiator (a controller, a smartphone app, or another device) sends a multicast message addressed to the group ID.
Only the devices that are members of that group will receive messages to or act on a message.
Using Group Communication reduces network traffic because a single multicast message can reach multiple devices.
This is more efficient than sending individual messages to each device.
It also enhances scalability as adding more devices to a group does not increase the number of messages required for group-wide commands.

.. note::

>    The communication reliability depends on the underlying networking technology.

Reviewer: Worth to mention this part which Kamil pointed:

> The group communication is by design not reliable, because there is no acks or responses, so we don't know if we reached all targets or not.

Reviewer 2: ```suggestion
   The group communication is by design not reliable, because it does not use any way of acknowledgement.
   Because of that the device that sends the message is not able to determine, whether it reached all devices belonging to the group. 
   The communication reliability depends on the underlying networking technology.
```

Group setup
***********

Devices are configured to be part of one or more groups.
Each group is identified by a unique group ID.
A device can be a member of multiple groups, and a group can include multiple devices.

Group Messaging
***************

When a command needs to be sent to all devices in a group, the command initiator (a controller, a smartphone app, or another device) sends a multicast message addressed to the group ID.
Only the devices that are members of that group will receive messages to or act on a message.
Using Group Communication reduces network traffic because a single multicast message can reach multiple devices.
This is more efficient than sending individual messages to each device.
It also enhances scalability as adding more devices to a group does not increase the number of messages required for group-wide commands.

.. note::

   Group Communication is not reliable by design.
   There are no acks or responses and the messages might not reach all targets.
   The communication reliability depends on the underlying networking technology.

>    For example, sleepy devices do not work well, as they will miss incoming messages, because the traffic is not buffered for them.

Reviewer 2: For example, sleepy devices should not be targeted in the group messaging, as they will likely miss incoming messages, because their radio is inactive for most of the time.
   The multicast messages are not buffered by their parents, and sent when the device is active, so they will not reach the destination.

>    For example, sleepy devices do not work well, as they will miss incoming messages, because the traffic is not buffered for them.

Reviewer 2: ```suggestion
   For example, sleepy devices should not be targeted in the group messaging, as they will likely miss incoming messages, because their radio is inactive for most of the time.
   The multicast messages are not buffered by their parents, and sent when the device is active, so they will not reach the destination.
```

Author: @kkasperczyk-no how about like this:
For example, sleepy devices should not be targeted in the group messaging, as they will likely miss incoming messages, because their radio is inactive for most of the time.
Multicast messages are sent when the parent is active and are not buffered, so they might not reach the destination.

Reviewer 2: It's not what I mean. The sleepy device is inactive, so it will not receive messages. Typically sleepy device's parent should buffer the messages for it and send them, once the sleepy device will be active. However it does not apply to multicast messages, these ones are not buffered.

Group Messaging
***************

When a command needs to be sent to all devices in a group, the command initiator (a controller, a smartphone app, or another device) sends a multicast message addressed to the group ID.
Only the devices that are members of that group will receive messages to or act on a message.
Using Group Communication reduces network traffic because a single multicast message can reach multiple devices.
This is more efficient than sending individual messages to each device.
It also enhances scalability as adding more devices to a group does not increase the number of messages required for group-wide commands.

.. note::

   Group Communication is by design not reliable, because it does not use any way of acknowledgment.
   Because of that, the device that sends the message cannot determine whether it reached all devices belonging to the group.
   Communication reliability depends on the underlying networking technology.
   For example, sleepy devices should not be targeted in the group messaging, as they will likely miss incoming messages, because their radio is inactive for most of the time.

>    Multicast messages are not buffered by their parents, and are sent when the device is active, so they will not reach the destination.

Reviewer 2: This is not right, they are not buffered, so they cannot be sent.
```suggestion
   Multicast messages are not buffered by their parents, so they will not reach the destination.
```

I have impression that I tried to introduce to many information and it mixed things up a bit :) Generally what I've meant is that parent sends buffered messages when the sleepy device is active. But in case of multicasts the messages are not buffered, so it won't work. Maybe it's too detailed to write all of these, so removing just part of the sentence above should work.
